### ISSUE TITLE
Make it worker friendly

### ISSUE THREAD

**Issue by laszbalo** [WebReflection/viperHTML#26]
Hi, I would like to use viperHTML inside a service worker, however it is not possible at the moment because of the runtime dependencies. It would be nice to have the templates compiled before they got shipped to the service worker, so that the service worker would not have to  incude an HTML parser, JS and CSS minifiers. It would be also nice to not depend on the node 'crypto' module. What I am envisioning is the following

```javascript
// src/templates.js
const text = 'Ugly Bugly'
const render = wire()
render`<h1>${text}</h1>`
render`<h1>${text}</h1>`
```
would be transpiled into
```javascript
// dist/__templates_internal.js
module.exports {
  hash1: {
    chunks: ['<h1>', '</h1>'],
    updates: [
      0, // this is the index of getUpdateForHTML function reference in the updateMappings array
     // [1, key] // index of updateBoolean, followed by argument 'key' for updateBoolean function
    ]  // this assumes, that there is a map in viperHTML, like : const updateMappings = [getUpdateForHTML, updateBoolean, etc..]
  }
}

// dist/templates.js
const TEMPLATES = require('./__templates_internal')

const text = 'Ugly Bugly'
const render = wire()
render(TEMPLATES.hash1.chunks, text)
render(TEMPLATES.hash1.chunks, text)
```
The above transpilation step would minify the JS and CSS, and assign the right type of updates for the interpolations. viperHTML however should be capable to map the template chunks to their respecitve updates array. What do you think about this?

**Comment by WebReflection** [comment]
viperHTML is for Nodejs. If you want to generate layouts in Service Workers use basicHTML.

You have all examples in here: https://github.com/WebReflection/hyperSW

**Comment by laszbalo** [comment]
Thanks, I did not know about basicHTML.

**Comment by laszbalo** [comment]
Checked out basicHTML, but it just does not seem to be an ideal solution, because of its massive file size. (my guess is that my bundler of choice,webpack, can not tree shake it because basicHTML uses CJS modules)

 After minification, it still weighs over 150Kb, more than what viperHTML weighs after removing the dependencies on csso and uglify.

Do you have any advice on how to bring the bundle size down?

**Comment by WebReflection** [comment]
> Do you have any advice on how to bring the bundle size down?

Use gzip?

<img width="663" alt="screen shot 2018-06-23 at 12 44 31" src="https://user-images.githubusercontent.com/85749/41808791-4db16642-76e3-11e8-9b1a-d0edc1f6cd5e.png">

**Comment by laszbalo** [comment]
Will gzip will help with the CPU time and memory consumption as well? :)

**Comment by WebReflection** [comment]
TL;DR viperHTML is for Node.js and it will never drop its needed dependencies to work in Node.js.

There is no DOM in a Worker so you need some DOM if you want to use the same logic hyperHTML uses. With viperHTML, the DOM tree is crawled by XML parsers, in hyperHTML it's based on native DOM parser or, in case of a worker, you can use basicHTML.

If you don't want any of these, you are in danger territory 'cause regex based parsing will easily fail in so many cases that I would not suggest you to go in that direction.

Without any recognition for special attributes, accessors, and everything else, you are also basically asking to make viperHTML. or hyperHTML, just a wrap for a template literal.

Well, you can do that too.

```js
function html(t) {
  for (var o = [t[0]], i = 1, l = arguments.length; i < l; i++) {
    o.push(arguments[i], t[i]);
  }
  return o.join('');
}
```

You can use that to produce any html you want but you are fully in charge of fixing all the issue you will have with such approach.

```js
html`<strong>Hello ${user}</strong>`;
```

Or, you can simply use basicHTML since it's in the Service Worker, downloaded once but always available offline, and it weights 45K compressed, which is nothing compared to React, as example, except it gives you infinite amount of offline pages.

**Comment by WebReflection** [comment]
> Will gzip will help with the CPU time and memory consumption as well? :)

are you speculating for the sake of it?

Service Worker is available only in modern Browsers that will run in modern hardware so unless you have real-world issues don't bother optimizing anything, it's not needed, I am a performance maniac myself and what I write is written to be performant by default.

**Comment by laszbalo** [comment]
> TL;DR viperHTML is for Node.js and it will never drop its needed dependencies to work in Node.js.

I never meant to drop dependencies, I was just curious what do you think about separating viperHTML into two logical parts:

1. first part is some kind of compiler/transpiler which would be able to create a JSON file which consists of the the sanitized chunks, and the update information for the interpolations. (basically calling transform at compile time [transform](https://github.com/WebReflection/viperHTML/blob/master/index.js#L218) returning JSON)
2. a modified version of viperHTML which would use the above JSON representation

so that someone could just bundle the second part when targeting web workers without needing to include the first.

I might be wrong on this but the second part seems easily doable. However, when it comes to the first part, the  the compiler would require to collect all the tagged template literals at compile time which would be called with the viperHTML context at run time. Since the dynamic nature of JS, I am not even sure that making assumptions about the context of a JS function will be running in makes sense at all.

> what I write is written to be performant by default.

Have no doubt about that. To be honest, I find your projects quite ingenious.

Being cautious about what to include in my bundle does not necessary constitute to premature optimization. If size was not matter, I would still be using React.

**Comment by WebReflection** [comment]
I don't think JSON, instead of XML/HTML, would make this project any better, but if you want JSON to create layouts then you can use the good old JSONML: http://www.jsonml.org

But like I've said, I've solved the issue already via basicHTML, and every mobile browser with not so great hardware didn't sweat a second with it.

There are no memory issues, and 45K gzipped are nothing compared to the rest of the Web.

Those 45K are also off the main thread so are literally not blocking, neither interfering with the regular loading page.

However, if you can use React on a Service Worker and you are happy with React, then go with React.

There are various Web projects that use React, which weights more than basicHTML and also is way more aggressive on RAM and CPU usage, that asking me to maintain or refactor from scratch something that already works so well on Node.js, makes little sense, IMO.

### As Summary
If you need a lightweight DOM with Custom Elements ability and DOM Level 0 methods, on a Worker, you can use basicHTML.

The day you will have real issues with that, you can think about a better alternative, since the usage of template literals, a standard JS feature compared to JSX, does not lock you into any library.

**Comment by WebReflection** [comment]
P.S. it's 43K here, if the compression is your issue, use a CDN such https://unpkg.com/basichtml
